include/deal.II/matrix_free/tensor_product_kernels.h

```diff
@@ -3994,6 +4183,279 @@ namespace internal
+      {
+        // test value z
+        test_value = dim > 2 ? value * shapes[i2][0][2] : value;
+
+        // Generate separate code with known loop bounds for the most common
+        // cases
+        if (n_shapes == 2)
+          do_apply_test_functions_xy_value<dim, 2, Number2, Number, add>(
+            values, shapes, test_value, n_shapes, i);
+        else if (n_shapes == 3)
+          do_apply_test_functions_xy_value<dim, 3, Number2, Number, add>(
+            values, shapes, test_value, n_shapes, i);
+        else if (n_shapes == 4)
+          do_apply_test_functions_xy_value<dim, 4, Number2, Number, add>(
+            values, shapes, test_value, n_shapes, i);
+        else if (n_shapes == 5)
+          do_apply_test_functions_xy_value<dim, 5, Number2, Number, add>(
+            values, shapes, test_value, n_shapes, i);
+        else if (n_shapes == 6)
+          do_apply_test_functions_xy_value<dim, 6, Number2, Number, add>(
+            values, shapes, test_value, n_shapes, i);
+        else
+          do_apply_test_functions_xy_value<dim, -1, Number2, Number, add>(
+            values, shapes, test_value, n_shapes, i);
+      }
+  }
+
+
+
+  /**
+   * Specializes @p integrate_tensor_product_value_shapes() for linear
+   * polynomials which massively reduces the necessary instructions.
+   */
+  template <int dim, typename Number, typename Number2, bool add>
+  inline void
+  integrate_add_tensor_product_value_linear(const unsigned int        n_shapes,
+                                            const Number2 &           value,
+                                            Number2 *                 values,
+                                            const Point<dim, Number> &p)
+  {
+    (void)n_shapes;
+    static_assert(dim >= 0 && dim <= 3, "Only dim=0,1,2,3 implemented");
+
+    AssertDimension(n_shapes, 2);
+
+    if (dim == 0)
+      {
+        if (add)
+          values[0] += value;
+        else
+          values[0] = value;
+      }
+    else if (dim == 1)
+      {
+        const auto x0 = 1. - p[0], x1 = p[0];
+
+        if (add)
+          {
+            values[0] += value * x0;
+            values[1] += value * x1;
+          }
+        else
+          {
+            values[0] = value * x0;
+            values[1] = value * x1;
+          }
+      }
+    else if (dim == 2)
+      {
+        const auto x0 = 1. - p[0], x1 = p[0], y0 = 1. - p[1], y1 = p[1];
+
+        const auto test_value_y0 = value * y0;
+        const auto test_value_y1 = value * y1;
+
+        if (add)
+          {
+            values[0] += x0 * test_value_y0;
+            values[1] += x1 * test_value_y0;
+            values[2] += x0 * test_value_y1;
+            values[3] += x1 * test_value_y1;
+          }
+        else
+          {
+            values[0] = x0 * test_value_y0;
+            values[1] = x1 * test_value_y0;
+            values[2] = x0 * test_value_y1;
+            values[3] = x1 * test_value_y1;
+          }
```

Comment:
As opposed to the evaluate function above, I do not see that the alternative evaluation, using the distributive law on `values[0] += (1 - p[0]) * test_value_y0` (transformed into `values[0] += test_value_y0; values[0] -= p[0] * test_value_y0;`) and `values[1] += p[0] * test_value_y0` would yield a performance benefit, at least not with FMA instructions. The reason is that we need to have the extra addition twice, whereas `x0` allows to combine them. On the other hand, the assignment version as well as `y0/y1` might benefit from the alternative form.

Reply:
I think we should try to optimize the kernels in follow-ups!